cli/commands/commands.go

```
@@ -542,22 +578,30 @@ var killServices = &cobra.Command{
 			return nil
 		}
 
-		_, err = connectToHub().StopServices(context.Background(), &idl.StopServicesRequest{})
-		if err != nil {
-			errCode := grpcStatus.Code(err)
-			errMsg := grpcStatus.Convert(err).Message()
-			// XXX: "transport is closing" is not documented but is needed to uniquely interpret codes.Unavailable
-			// https://github.com/grpc/grpc/blob/v1.24.0/doc/statuscodes.md
-			if errCode != codes.Unavailable || errMsg != "transport is closing" {
-				return err
-			}
-			return nil
+		err, done := stopAllServices()
+		if done {
+			return err
 		}
 
 		return nil
 	},
 }
 
+func stopAllServices() (error, bool) {
+	_, err := connectToHub().StopServices(context.Background(), &idl.StopServicesRequest{})
+	if err != nil {
+		errCode := grpcStatus.Code(err)
+		errMsg := grpcStatus.Convert(err).Message()
+		// XXX: "transport is closing" is not documented but is needed to uniquely interpret codes.Unavailable
+		// https://github.com/grpc/grpc/blob/v1.24.0/doc/statuscodes.md
+		if errCode != codes.Unavailable || errMsg != "transport is closing" {
+			return err, true
+		}
+		return nil, true
```

could you explain why err is marked as nil here?

---

> could you explain why err is marked as nil here?

Its nil since we want ignore errors that are caused by the gRPC sever shutting down (which is Unavailable - "transport is closing"). That is, we only want to return errors that are not of that type. See https://github.com/grpc/grpc/blob/v1.24.0/doc/statuscodes.md


What I don't yet fully understand is the need for the boolean.

---

I got rid of the boolean.